File: NeetCode/Arrays_and_Hashing/First_Missing_Positive/first__missing__positive.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)

        def hop(val):
            while nums[val - 1] != val:
                curr = nums[val - 1]
                nums[val - 1] = val
                if curr <= 0 or curr > n + 1:
                    break
                val = curr

        for index in range(len(nums)):
            if nums[index] <= 0 or nums[index] > n:
                nums[index] = 0
                continue
            hop(nums[index])

        for index in range(len(nums)):
            if nums[index] != index + 1:
                return index + 1
        return n + 1


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 2, 4]
    print(sol.firstMissingPositive(nums))
    nums = [7, 8, 9, 10, 11]
    print(sol.firstMissingPositive(nums))
    nums = [1, 2, 4, 5, 6, 3, 1]
    print(sol.firstMissingPositive(nums))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/First_Missing_Positive/First_Missing_Positive.cpp
// First_Missing_Positive.cpp
//
// Compile locally with:
// clang++ First_Missing_Positive.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {
        int n = nums.size();
        for (int index = 0; index < n; index++)
        {
            if (nums[index] <= 0 || nums[index] > n)
            {
                nums[index] = 0;
                continue;
            }
            hop(nums[index], nums);
        }
        for (int idx = 0; idx < n; idx++)
        {
            if (nums[idx] != idx + 1)
            {
                return idx + 1;
            }
        }
        return n + 1;
    }
    void hop(int value, vector<int> &nums)
    {
        while (nums[value - 1] != value)
        {
            int curr = nums[value - 1];
            nums[value - 1] = value;
            if (curr <= 0 || curr > nums.size())
            {
                break;
            }
            value = curr;
        }
    }
};

int main()
{
    cout << "Running First_Missing_Positive.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
